stub/src/main/java/io/grpc/stub/MetadataUtils.java

```diff
@@ -143,4 +148,56 @@ public void onClose(Status status, Metadata trailers) {
       }
     }
   }
+
+  /**
+   * Returns a ServerInterceptor that attaches a given set of headers to every response.
+   *
+   * @param extraHeaders the headers to be added to each response. Caller gives up ownership.
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/11462")
+  public static ServerInterceptor newAttachHeadersServerInterceptor(Metadata extraHeaders) {
+    return new MetadataAttachingServerInterceptor(extraHeaders);
+  }
+
+  private static final class MetadataAttachingServerInterceptor implements ServerInterceptor {
+
+    private final Metadata extraHeaders;
+
+    MetadataAttachingServerInterceptor(Metadata extraHeaders) {
+      this.extraHeaders = extraHeaders;
+    }
+
+    @Override
+    public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+        ServerCall<ReqT, RespT> call, Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+      return next.startCall(new HeaderAttachingServerCall<>(call), headers);
+    }
+
+    final class HeaderAttachingServerCall<ReqT, RespT>
+        extends SimpleForwardingServerCall<ReqT, RespT> {
+      boolean headersSent;
+
+      HeaderAttachingServerCall(ServerCall<ReqT, RespT> delegate) {
+        super(delegate);
+      }
+
+      @Override
+      public void sendHeaders(Metadata headers) {
+        headers.merge(extraHeaders);
+        headersSent = true;
+        super.sendHeaders(headers);
+      }
+
+      @Override
+      public void close(Status status, Metadata trailers) {
+        if (!headersSent) {
+          // It isn't too late to call sendHeaders(): !headersSent implies that it hasn't been
+          // called yet (obviously). But it also implies that no messages have been sent, because
+          // sendMessage() *requires* a preceding call to sendHeaders().
+          sendHeaders(new Metadata());
```

Comment:
Oh, you're forcing headers. I don't think we want to do that generally. Maybe that is what you need, but I think it then stops being a general util.

Trailers-only is important to retries, as it delivers the status code without committing.

Reply:
I didn't know that - fixed and thanks!